v1/tarball/write.go

```diff
@@ -32,14 +32,24 @@ type WriteOptions struct {
 	// TODO(mattmoor): Whether to store things compressed?
 }
 
-// Write saves the image as the given tag in a tarball at the given path.
-func Write(p string, tag name.Tag, img v1.Image, wo *WriteOptions) error {
-	// Write in the compressed format.
-	// This is a tarball, on-disk, with:
-	// One manifest.json file at the top level containing information about several images.
-	// One file for each layer, named after the layer's SHA.
-	// One file for the config blob, named after its SHA.
+// Write writes the image at the given tag to a tarball,
+// and returns it through a reader to be consumed
+func Write(tag name.Tag, img v1.Image, wo *WriteOptions) io.ReadCloser {
```

Comment:
It's weird to me that something called `Write` returns a `ReadCloser`. After spending a few minutes trying to come up with a name that made more sense, I think I realized the issue.

This would be a lot cleaner if `tarball.Write` was instead:
```go
func Write(w io.Writer, tag name.Tag, img v1.Image, wo *WriteOptions) error
```

(Hey look, that's `tarball.write` ;P)

Then `tarball.WriteToFile` could just be sugar to create the file for you:
```go
func WriteToFile(p string, tag name.Tag, img v1.Image, wo *WriteOptions) error {
	f, err := os.Create(p)
	if err != nil {
		return err
	}
	defer f.Close()

	return Write(tag, img, wo, f)
}
```

And then you can move the `io.Pipe` from `tarball.Write` into `daemon.Write`, which would just pipe `tarball.Write` into `cli.ImageLoad`.

@ImJasonH thoughts? I think this would be a bit more idiomatic/composable.

Reply:
Yes! I was bumping up on that too and just didn't think about it enough. I think we could probably port that same change to `mutate.Extract` too (why is `Extract` in `mutate`?)

But yeah that's definitely a better API

Reply:
+1 to giving `mutate.Extract` the same treatment.

Disclaimer: the rest of this comment is unrelated to this PR.

tl;dr I think we need `fs.Write` and `fs.Image` (or `fs.Layer`)

> (why is Extract in mutate?)

I'll bring up [my comment](https://github.com/google/go-containerregistry/pull/99#issuecomment-383757347) again -- if we can change `mutate.Extract` to:

```go
func Flatten(v1.Image) (v1.Image, error)
```

... I think it makes sense to leave it in `mutate`. Otherwise, I agree it doesn't really fit the style of mutations in there (maybe it belongs in `v1util` or `tarball`? or a new `fs.Write`?). Unfortunately, the change I want makes things a bit complicated and potentially a lot slower...

I'm wondering if we can abuse this to get what I want without much code... e.g.:

```go
func Flatten(i v1.Image) (v1.Image, error) {
  // pseudo-go, will likely need lots of plumbing to avoid excessive buffering
  return mutate.AppendLayers(empty.Image(), layer.Uncompressed(mutate.Extract(i)))
}
```

Then implementing `{remote,tarball,daemon} | flatten | {remote,tarball,daemon}` is trivial :)

Reply:
Thanks for the feedback here, this is definitely a better construction. Passing in a writer rather than returning a reader puts the onus on the caller to do the IO, which I think is a lot cleaner.

re `Extract` being in `mutate`: I originally had this in `tarball`, but wasn't sure it made sense in there since it's not technically limited to `tarball`s (though they do get converted before doing the actual extraction).

The actual motivation for this was twofold: I wanted to have a cleaner way of doing filesystem comparisons in container-diff (extracting the flattened fs is an obvious choice), but then I also wanted to implement `Flatten` in its original state because it's cool :) 

I tried to make it as extensible as possible, and I guess `mutate` was the only neutral package that made sense. I'm fine moving it around though.